--list-modules CLI flag produces segfault

The `--list-modules` CLI flag produces a segfault:

```
› make && ./hotstuff --list-modules
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xe8 pc=0x70ca1c]

goroutine 1 [running]:
github.com/relab/hotstuff/modules.ListModules()
	/home/michael/dev/go/hotstuff/modules/registry.go:89 +0x1dc
github.com/relab/hotstuff/internal/cli.glob..func1(0x11be760?, {0xc1f2a6?, 0x1?, 0x1?})
	/home/michael/dev/go/hotstuff/internal/cli/root.go:37 +0x34
github.com/spf13/cobra.(*Command).execute(0x11be760, {0xc0001a8010, 0x1, 0x1})
	/home/michael/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0x11be760)
	/home/michael/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/home/michael/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
github.com/relab/hotstuff/internal/cli.Execute()
	/home/michael/dev/go/hotstuff/internal/cli/root.go:52 +0x25
main.main()
	/home/michael/dev/go/hotstuff/cmd/hotstuff/main.go:7 +0x17
```
Following the stack trace shows it must be due to the `t` variable, a key of the `byInterface` map, being `<nil>`: https://github.com/relab/hotstuff/blob/master/modules/registry.go#L89

The queried datastructure (`byInterface`) is only filled within here, where the offending map key is called `moduleType`: https://github.com/relab/hotstuff/blob/master/modules/registry.go#L31-L34

`moduleType` is set in the same function as `moduleType := reflect.TypeOf(t)`. Here, `var t T` is (the zero value of) a variable of type `T`. `T` is seemingly the (generic) return type of the constructor of the module which is being registered.

Which brings us to the likely cause. The return type of at least some of the registered modules' constructors is an interface, e.g here: https://github.com/relab/hotstuff/blob/master/consensus/byzantine/byzantine.go#L11

Thus `var t T` has the zero value of an interface, which is `<nil>`. `moduleType := reflect.TypeOf(t)` [returns `nil` as `t` is a `nil` interface](https://pkg.go.dev/reflect#TypeOf). This ends up in the `byInterface` map. And eventually it blows up when one tries to access it as something of type `reflect.Type`.

I don't understand why you seemingly keep modules grouped by their constructor's return value in the `byInterface` data structure, and why you want to print them that way in the `--list-modules` call, so am hesitant to recommend a fix.

--- a/modules/registry.go
+++ b/modules/registry.go
@@ -14,10 +14,11 @@ var (
 
 // RegisterModule registers a constructor for a module implementation with the specified name.
 // For example:
-//  RegisterModule("chainedhotstuff", func() consensus.Rules { return chainedhotstuff.New() })
+//
+//	RegisterModule("chainedhotstuff", func() consensus.Rules { return chainedhotstuff.New() })
 func RegisterModule[T any](name string, constructor func() T) {
 	var t T
-	moduleType := reflect.TypeOf(t)
+	moduleType := reflect.TypeOf(&t).Elem()
 
 	registryMut.Lock()
 	defer registryMut.Unlock()
@@ -39,7 +40,8 @@ func RegisterModule[T any](name string, constructor func() T) {
 // GetModule constructs a new instance of the module with the specified name.
 // GetModule returns true if the module is found, false otherwise.
 // For example:
-//  rules, ok := GetModule[consensus.Rules]("chainedhotstuff")
+//
+//	rules, ok := GetModule[consensus.Rules]("chainedhotstuff")
 func GetModule[T any](name string) (out T, ok bool) {
 	targetType := reflect.TypeOf(out)
 

--- a/modules/registry.go
+++ b/modules/registry.go
@@ -43,7 +43,7 @@ func RegisterModule[T any](name string, constructor func() T) {
 //
 //	rules, ok := GetModule[consensus.Rules]("chainedhotstuff")
 func GetModule[T any](name string) (out T, ok bool) {
-	targetType := reflect.TypeOf(out)
+	targetType := reflect.TypeOf(&out).Elem()
 
 	registryMut.Lock()
 	defer registryMut.Unlock()

--- a/modules/registry_test.go
+++ b/modules/registry_test.go
@@ -1,16 +1,34 @@
 package modules_test
 
 import (
+	"fmt"
+	"reflect"
 	"testing"
 
 	"github.com/relab/hotstuff/modules"
+	"golang.org/x/exp/slices"
 )
 
-func TestModuleRegistry(t *testing.T) {
+func init() {
+	// register some modules
 	modules.RegisterModule("frobulator", func() moduleIface {
 		return module{}
 	})
 
+	modules.RegisterModule("fizzbuzz", func() moduleIface {
+		return fizzbuzz{}
+	})
+
+	modules.RegisterModule("fizzer", func() fizzer {
+		return fizzbuzz{}
+	})
+
+	modules.RegisterModule("buzzer", func() buzzer {
+		return fizzbuzz{}
+	})
+}
+
+func TestModuleRegistry(t *testing.T) {
 	frobulator, ok := modules.GetModule[moduleIface]("frobulator")
 	if !ok {
 		t.Fatal("module was not found")
@@ -23,6 +41,32 @@ func TestModuleRegistry(t *testing.T) {
 	}
 }
 
+func TestListModules(t *testing.T) {
+	getName := func(t reflect.Type) string {
+		return fmt.Sprintf("(%s).%s", t.PkgPath(), t.Name())
+	}
+
+	want := map[string][]string{
+		getName(reflect.TypeOf((*moduleIface)(nil)).Elem()): {"frobulator", "fizzbuzz"},
+		getName(reflect.TypeOf((*fizzer)(nil)).Elem()):      {"fizzer"},
+		getName(reflect.TypeOf((*buzzer)(nil)).Elem()):      {"buzzer"},
+	}
+
+	list := modules.ListModules()
+
+	for iface, names := range want {
+		if got, ok := list[iface]; !ok {
+			t.Errorf("expected interface '%s' to be in list of modules", iface)
+		} else {
+			for _, name := range names {
+				if !slices.Contains(got, name) {
+					t.Errorf("expected '%s' to be in list of '%s' implementations", name, iface)
+				}
+			}
+		}
+	}
+}
+
 type moduleIface interface {
 	frobulate(i *int)
 }
@@ -32,3 +76,19 @@ type module struct{}
 func (module) frobulate(i *int) {
 	*i++
 }
+
+type fizzer interface {
+	fizz(i int)
+}
+
+type buzzer interface {
+	buzz(i int)
+}
+
+type fizzbuzz struct{}
+
+func (f fizzbuzz) fizz(_ int) { panic("not implemented") }
+
+func (f fizzbuzz) buzz(_ int) { panic("not implemented") }
+
+func (f fizzbuzz) frobulate(_ *int) { panic("not implemented") }

